server/src/main/java/org/elasticsearch/index/shard/RemoveCorruptedShardDataCommand.java

```diff
@@ -0,0 +1,584 @@
+            throw new ElasticsearchException("index directory [" + indexPath + "], must exist and be a directory");
+        }
+
+        final IndexMetaData indexMetaData =
+            IndexMetaData.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, shardParentPath);
+        final Index index = indexMetaData.getIndex();
+        final ShardId shardId;
+        final String fileName = path.getFileName().toString();
+        if (Files.isDirectory(path) && fileName.chars().allMatch(Character::isDigit)) {
+            int id = Integer.parseInt(fileName);
+            shardId = new ShardId(index, id);
+        } else {
+            throw new ElasticsearchException("Unable to resolve shard id from " + path.toString());
+        }
+
+        final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings);
+
+        return resolveShardPath(environment,
+            (nodeLockId, nodePath) -> {
+                final Path shardPathLocation = nodePath.resolve(shardId);
+                final ShardPath shardPath = ShardPath.loadShardPath(logger, shardId, indexSettings, new Path[]{shardPathLocation},
+                    nodeLockId, nodePath.path);
+                if (shardPath != null && shardPath.resolveIndex().equals(indexPath)) {
+                    return shardPath;
+                }
+                return null;
+            },
+            () -> new ElasticsearchException("Unable to resolve shard path for path " + path.toString()));
+    }
+
+    private ShardPath resolveShardPath(final Environment environment,
+                                       final CheckedBiFunction<Integer, NodeEnvironment.NodePath, ShardPath, IOException> function,
+                                       final Supplier<ElasticsearchException> exceptionSupplier)
+        throws IOException, UserException {
+        // have to iterate over possibleLockId as NodeEnvironment - but fail if node owns lock
+        final Settings settings = environment.settings();
+        final Path[] dataFiles = environment.dataFiles();
+        // try to resolve shard path in case of multi-node layout per environment
+        final int maxLocalStorageNodes = NodeEnvironment.MAX_LOCAL_STORAGE_NODES_SETTING.get(settings);
+        for (int dirIndex = 0; dirIndex < dataFiles.length; dirIndex++) {
+            final Path dataDir = dataFiles[dirIndex];
+            for (int possibleLockId = 0; possibleLockId < maxLocalStorageNodes; possibleLockId++) {
+                final Path dir = NodeEnvironment.resolveNodePath(dataDir, possibleLockId);
+                if (Files.exists(dir) == false) {
+                    // assume that we do not have gaps in nodes like
+                    break;
+                }
+                try (Directory nodeDir = FSDirectory.open(dir, NativeFSLockFactory.INSTANCE);
+                     Lock lock = nodeDir.obtainLock(NodeEnvironment.NODE_LOCK_FILENAME)) {
```

Comment:
I think we're using this lock wrongly here for two reasons:

1. we drop it before we resolve the shard path
2. we drop it and reacquire it before actually fixing anything.

I think we should obtain this lock once, do all our stuff, and only then drop it. If we obtain the lock and then discover that this is the wrong node then drop it and move onto the next node.

Reply:
agreed. reworked